runtime/src/main/cpp/Memory.cpp

```diff
@@ -451,6 +451,10 @@ struct MemoryState {
   uint64_t allocSinceLastGcThreshold;
 #endif // USE_GC
 
+  // This map is expected to be small, std::map consumes less memory than
+  // std::unordered_map and is just as efficient.
+  KStdOrderedMap<ObjHeader**, ObjHeader*> initializingSingletons;
```

Comment:
Does using stack-like collection here has any effect on the performance?

Reply:
It does, actually. It outperforms `std::map` on benchmarks from https://github.com/JetBrains/kotlin-native/pull/4018 And compared to the current implementation: it's 5% faster on a single initialization, and 10% slower on a nested initialization (with depth 500). I think, it's a win overall, since deeply nested initialization should be very rare.